### PR TITLE
Implementing shake_128 and shake_256 in the hashlib module

### PR DESCRIPTION
**Pull request by haxelion**
I implemented support for hashlib shake_128 and shake_256.
Going through the code I made some refactoring to better match CPython behavior.

There is one test I had to disable however. The test was enabled but didn't execute any code because shake implementation were missing: https://github.com/RustPython/RustPython/blob/a7c985685146ca401011eb0c25c0b2805a767ce7/Lib/test/test_hashlib.py#L311-L312

But also I'm not sure we want to match openssl length size limitations in our implementation.

### CODE REVIEW

File: Lib/test/test_hashlib.py [hunk 1]
```diff
@@ -303,6 +300,8 @@ def test_hexdigest(self):
                 self.assertIsInstance(h.digest(), bytes)
                 self.assertEqual(hexstr(h.digest()), h.hexdigest())
 
+    # TODO: RUSTPYTHON
+    @unittest.skip
```
Comment (DimitrisJim) [inline Lib/test/test_hashlib.py:304]: does this crash the interpreter or could it just be an expected failure?

File: stdlib/src/hashlib.rs [hunk 2]
```diff
@@ -45,17 +55,33 @@ mod hashlib {
...
+            }
+        }
+    }
+
+    #[derive(FromArgs)]
+    #[allow(unused)]
+    struct XofDigestArgs {
+        #[pyarg(positional)]
+        length: usize,
+    }
```
Comment (youknowone) [inline stdlib/src/hashlib.rs:72]:
```suggestion
        length: isize,
    }

    impl XofDigestArgs {
        fn length(&self, vm: &VirtualMachine) -> PyResult<usize> {
            if length < 0 {
                return vm.new_value_error("length must be non-negative".to_owned());
            }
            Ok(length as usize)
        }
    }
```
the skipped test may be fixed with this

File: Lib/test/test_hashlib.py (hunk 1 above)
Comment (haxelion) [reply Lib/test/test_hashlib.py:304]: Looking at this test, I think I know how to fix it ... wait a second 🙂

Comment (haxelion) [reply Lib/test/test_hashlib.py:304]: Well, it's possible but probably worth a separate PR.
The test doesn't crash the interpreter so I'll mark it as expectedFailure.

File: stdlib/src/hashlib.rs (hunk 2 above)
Comment (haxelion) [reply stdlib/src/hashlib.rs:72]: Ok let me test this later 👍

**Comment by youknowone** [comment]
Looks great. Fixing the isize part will make the patch perfect.

**Comment by haxelion** [comment]
@youknowone So I tested it and it does pass the check at line 317:
https://github.com/RustPython/RustPython/blob/a7c985685146ca401011eb0c25c0b2805a767ce7/Lib/test/test_hashlib.py#L317

The issue is that instead we will fail on these lines (after a lengthy computation): https://github.com/RustPython/RustPython/blob/a7c985685146ca401011eb0c25c0b2805a767ce7/Lib/test/test_hashlib.py#L319-L320

Because we do support those sizes:
https://github.com/RustPython/RustPython/blob/a7c985685146ca401011eb0c25c0b2805a767ce7/Lib/test/test_hashlib.py#L308

So if we go back to CPython, they have two implementations:

- A fallback implementation which does not support those larger sizes: https://github.com/python/cpython/blob/568fc0dee42a353f327b059a48f97c911de904b3/Modules/_sha3/sha3module.c#L455
- An default  (if the library is found on the system) OpenSSL implementation which does support those sizes and skip the test with this check: https://github.com/RustPython/RustPython/blob/a7c985685146ca401011eb0c25c0b2805a767ce7/Lib/test/test_hashlib.py#L313-L315

So there are basically two solutions:

1. Adhere to the fallback implementation and only accept an `unsigned long`.
2. Copy the module structure of the CPython OpenSSL implementation so that we can skip the test.

I'm unsure which solution you would prefer.

**Comment by youknowone** [comment]
The second one is preferred, but it doesn't sound like a part to cover in this PR anymore. 
Would you finish them after only changing the `usize` to `isize`?

**Comment by youknowone** [comment]
I am not sure but second approach also will not be that hard.
If you'd like to try,

adding
```rust
#[pyclass]
struct HASH;
```
and changing `PyHashser` to
```rust
#[pyclass(module = "hashlib", name = "hasher", base = "HASH")]
```
will be a trial

**Comment by haxelion** [comment]
@youknowone So I already tried this change but it's not enough. The classes need to be in a separate `_hashlib` module for this to work ... I tried the change but it became a much larger PR so I gave up.

I will push the change with `isize`.

**Review by youknowone — APPROVED** [review]
Thank you!

**Comment by youknowone** [comment]
I got what you meant. both CPython's `hashilb` and `test_hashlib` heavily suppose `_hashlib` is openssl implementation.

**Pull request merged**